pubsub-mapped-api/src/main/java/com/google/pubsub/clients/consumer/KafkaConsumer.java

```diff
@@ -0,0 +1,633 @@
+    checkPatternSubscribePreconditions(pattern);
+
+    List<Topic> existingTopics = getPubsubExistionTopics();
+
+    List<String> matchingTopics = new ArrayList<>();
+
+    for (Topic topic: existingTopics) {
+      String topicName = topic.getName().substring(TOPIC_PREFIX.length(), topic.getName().length());
+      Matcher m = pattern.matcher(topicName);
+      if (m.matches())
+        matchingTopics.add(topicName);
+    }
+    subscribe(matchingTopics);
+
+    log.debug("Subscribed to pattern: {}", pattern);
+  }
+
+  private List<Topic> getPubsubExistionTopics() {
+    ListTopicsResponse listTopicsResponse = publisherBlockingStub
+        .listTopics(ListTopicsRequest.newBuilder()
+            .setProject(GOOGLE_CLOUD_PROJECT)
+            .build());
+
+    return listTopicsResponse.getTopicsList();
+  }
+
+  private void checkPatternSubscribePreconditions(Pattern pattern) {
+    Preconditions.checkArgument(pattern != null,
+        "Topic pattern to subscribe to cannot be null");
+  }
+
+  @Override
+  public void unsubscribe() {
+    unsubscribe(subscriptions.values());
+    subscriptions = ImmutableMap.of();
+  }
+
+  private void unsubscribe(Collection<Subscription> subscriptions) {
+    List<ListenableFuture<Empty>> listenableFutures = new ArrayList<>();
+    for (Subscription s: subscriptions) {
+      ListenableFuture<Empty> emptyListenableFuture = subscriberFutureStub
+          .deleteSubscription(DeleteSubscriptionRequest.newBuilder()
+              .setSubscription(s.getName()).build());
+
+      listenableFutures.add(emptyListenableFuture);
+    }
+
+    ListenableFuture<List<Empty>> listListenableFuture = Futures.allAsList(listenableFutures);
+    Futures.addCallback(listListenableFuture, new SubscribeCallback());
+  }
+
+  @Override
+  public ConsumerRecords<K, V> poll(long timeout) {
+    checkPollPreconditions(timeout);
+
+    Map<TopicPartition, List<ConsumerRecord<K, V>>> pollRecords = new HashMap<>();
+    List<ResponseData<PullResponse>> responsesFutures = deputePubsubPulls(timeout);
+
+    List<AcknowledgeRequest> acknowledgeRequests = new ArrayList<>();
+
+    for (ResponseData<PullResponse> pollData : responsesFutures) {
+      try {
+        TopicPartition topicPartition = new TopicPartition(pollData.getTopicName(),
+            DEFAULT_PARTITION);
+
+        PullResponse pulled = pollData.getRequestListenableFuture().get();
+
+        List<ConsumerRecord<K, V>> subscriptionRecords = mapToConsumerRecords(pollData,
+            pulled);
+
+        if (!pulled.getReceivedMessagesList().isEmpty()) {
+          AcknowledgeRequest acknowledgeRequest = getAcknowledgeRequest(
+              pollData.getSubscriptionFullName(), pulled);
+          acknowledgeRequests.add(acknowledgeRequest);
+        }
+
+        pollRecords.put(topicPartition, subscriptionRecords);
+
+        //if all messages were pulled correctly, send ACK
+        acknowledgeMessages(acknowledgeRequests);
```

Comment:
This means we are acking before we have delivered to the user. Kafka allows you to configure this, correct? Either the Kafka client automatically commits at a time interval or one calls commit[Async] to do so, right? I don't think we want to ack at this point.